src/main/java/io/fabric8/jenkins/openshiftsync/PipelineJobListener.java

```diff
@@ -16,126 +16,371 @@
+						.withName(buildConfigName).get();
+				if (buildConfig != null) {
+					try {
+						getAuthenticatedOpenShiftClient().buildConfigs().inNamespace(namespace)
+								.withName(buildConfigName).delete();
+					} catch (KubernetesClientException e) {
+						if (HTTP_NOT_FOUND != e.getCode()) {
+							logger.log(Level.WARNING, "Failed to delete BuildConfig in namespace: " + namespace
+									+ " for name: " + buildConfigName, e);
+						}
+					} catch (Exception e) {
+						logger.log(Level.WARNING, "Failed to delete BuildConfig in namespace: " + namespace
+								+ " for name: " + buildConfigName, e);
+					} finally {
+						removeJobWithBuildConfig(buildConfig);
+					}
+				}
+			}
+		}
+	}
 
-  @Override
-  public void onDeleted(Item item) {
-    super.onDeleted(item);
-    if (item instanceof WorkflowJob) {
-      WorkflowJob job = (WorkflowJob) item;
-      if (job.getProperty(BuildConfigProjectProperty.class) != null
-        && isNotBlank(job.getProperty(BuildConfigProjectProperty.class).getNamespace())
-        && isNotBlank(job.getProperty(BuildConfigProjectProperty.class).getName())) {
-
-        NamespaceName buildName = OpenShiftUtils.buildConfigNameFromJenkinsJobName(job.getName(), job.getProperty(BuildConfigProjectProperty.class).getNamespace());
-        logger.info("Deleting BuildConfig " + buildName);
-
-        String namespace = buildName.getNamespace();
-        String buildConfigName = buildName.getName();
-        BuildConfig buildConfig = getOpenShiftClient().buildConfigs().inNamespace(namespace).withName(buildConfigName).get();
-        if (buildConfig != null) {
-          try {
-            getOpenShiftClient().buildConfigs().inNamespace(namespace).withName(buildConfigName).delete();
-          } catch (KubernetesClientException e) {
-            if (HTTP_NOT_FOUND != e.getCode()) {
-              logger.log(Level.WARNING, "Failed to delete BuildConfig in namespace: " + namespace + " for name: " + buildConfigName, e);
+	public void upsertItem(Item item) {
+		if (item instanceof WorkflowJob) {
+			upsertWorkflowJob((WorkflowJob) item);
+		} else if (item instanceof ItemGroup) {
+			upsertItemGroup((ItemGroup) item);
+		} else if (item instanceof Job) {
+      upsertJob((Job) item);
+    }
+	}
+
+  private void upsertJob(AbstractItem job) {
+    BuildConfigProjectProperty property = ConfigMapToJobMap.getOrFindProperty((Job) job);
+    if (property != null) {
+      String newConfigXml = null;
+      XmlFile configFile = job.getConfigFile();
+      if (configFile != null) {
+        try {
+          newConfigXml = configFile.asString();
+        } catch (IOException e) {
+          logger.log(Level.WARNING, "Failed to get Job " + job.getName() + " config.xml due to " + e, e);
+          return;
+        }
+      }
+      if (newConfigXml == null) {
+        return;
+      }
+      // lets update the ConfigMap if the config.xml has changed
```

Comment:
Updating the config map based on jenkins UI changes might be something we want to wrapper with an on/off switch as well, even if we enable job creation from config maps ... something to discuss in more detail when you all start submitting PRs against upstream

Preemptively wrapping your new features in such a way will probably make merging into upstream faster ... i.e. even if we are not 100% sure of something, we can merge it, turn it off by default in the openshift jenkins image, and then sort out the details afterward.